core/src/main/java/org/apache/druid/data/input/impl/HttpEntity.java

```diff
@@ -80,23 +80,30 @@ protected String getPath()
     return t -> t instanceof IOException;
   }
 
-  public static InputStream openInputStream(URI object, String userName, PasswordProvider passwordProvider, long offset)
-      throws IOException
-  {
-    final URLConnection urlConnection = object.toURL().openConnection();
+  private static void addAuthHeader(URLConnection urlConnection, String userName, PasswordProvider passwordProvider){
     if (!Strings.isNullOrEmpty(userName) && passwordProvider != null) {
       String userPass = userName + ":" + passwordProvider.getPassword();
       String basicAuthString = "Basic " + Base64.getEncoder().encodeToString(StringUtils.toUtf8(userPass));
       urlConnection.setRequestProperty("Authorization", basicAuthString);
     }
+  }
+
+  public static InputStream openInputStream(URI object, String userName, PasswordProvider passwordProvider, long offset)
+      throws IOException
+  {
+    final URLConnection urlConnection = object.toURL().openConnection();
+    addAuthHeader(urlConnection, userName, passwordProvider);
     final String acceptRanges = urlConnection.getHeaderField(HttpHeaders.ACCEPT_RANGES);
     final boolean withRanges = "bytes".equalsIgnoreCase(acceptRanges);
     if (withRanges && offset > 0) {
       // Set header for range request.
       // Since we need to set only the start offset, the header is "bytes=<range-start>-".
       // See https://tools.ietf.org/html/rfc7233#section-2.1
-      urlConnection.addRequestProperty(HttpHeaders.RANGE, StringUtils.format("bytes=%d-", offset));
-      return urlConnection.getInputStream();
+      urlConnection.getInputStream().close();
+      final URLConnection newUrlConnection = object.toURL().openConnection();
+      addAuthHeader(newUrlConnection, userName, passwordProvider);
+      newUrlConnection.addRequestProperty(HttpHeaders.RANGE, StringUtils.format("bytes=%d-", offset));
```

Comment:
The above line seems to add the offset as an HTTP header. It is more common to add the offset as a query parameter: `http://server/get/offset=20`. Seems that the code already uses headers. Just pointing out that that is an unusual approach.

Reply:
This offset is being used to resume a file download from an external server.

If an external server supports accept-ranges in the header, we are specifying the offset in the header itself so that we can open a stream which reads only after the offset rather than having to skip them in the input stream itself.